README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Redis components for Golang

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The module for working with a Redis database of type key-value contains components for working with cache and locks: RedisCache and RedisLock.

The module contains the following packages:

- **Build** - factory default
- **Cache** - Redis Cache Components
- **Lock** - components of working with locks in Redis

<a name="links"></a> Quick links:

* [Configuration](http://docs.pipservices.org/concepts/configuration/)
* [API Reference](https://godoc.org/github.com/pip-services3-gox/pip-services3-redis-gox/)
* [Change Log](CHANGELOG.md)
* [Get Help](http://docs.pipservices.org/get_help/)
* [Contribute](http://docs.pipservices.org/contribute/)

## Use

Get the package from the Github repository:
```bash
go get -u github.com/pip-services3-gox/pip-services3-redis-gox@latest
```

## Develop

For development you shall install the following prerequisites:
* Golang v1.18+
* Visual Studio Code or another IDE of your choice
* Docker
* Git

Run automated tests:
```bash
go test -v ./test/...
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized test as:
```bash
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov**, **Levichev Dmitry** and **Danil Prisiazhnyi**.

The documentation is written by:
- **Levichev Dmitry**

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Redis components for Golang Changelog

## <a name="1.0.0"></a> 1.0.0 (2022-07-10) 

### Features
* **build** default factory components
* **cache** Redis Cache Components
* **lock** components of working with locks in Redis



test/fixture/LockFixture.go
package test_fixture

import (
	"context"
	"testing"

	clock "github.com/pip-services3-gox/pip-services3-components-gox/lock"
	"github.com/stretchr/testify/assert"
)

const (
	LOCK1 string = "lock_1"
	LOCK2 string = "lock_2"
	LOCK3 string = "lock_3"
)

type LockFixture struct {
	lock clock.ILock
}

func NewLockFixture(lock clock.ILock) *LockFixture {
	c := LockFixture{}
	c.lock = lock
	return &c
}

func (c *LockFixture) TestTryAcquireLock(t *testing.T) {
	ctx := context.Background()

	// Try to acquire lock for the first time
	result, err := c.lock.TryAcquireLock(ctx, "", LOCK1, 3000)
	assert.Nil(t, err)
	assert.True(t, result)

	// Try to acquire lock for the second time
	result, err = c.lock.TryAcquireLock(ctx, "", LOCK1, 3000)
	assert.Nil(t, err)
	assert.False(t, result)

	// Release the lock
	err = c.lock.ReleaseLock(ctx, "", LOCK1)
	assert.Nil(t, err)

	// Try to acquire lock for the third time
	result, err = c.lock.TryAcquireLock(ctx, "", LOCK1, 3000)
	assert.Nil(t, err)
	assert.True(t, result)

	c.lock.ReleaseLock(ctx, "", LOCK1)
}

func (c *LockFixture) TestAcquireLock(t *testing.T) {
	ctx := context.Background()

	// Acquire lock for the first time
	err := c.lock.AcquireLock(ctx, "", LOCK2, 3000, 1000)
	assert.Nil(t, err)

	// Acquire lock for the second time
	err = c.lock.AcquireLock(ctx, "", LOCK2, 3000, 1000)
	assert.NotNil(t, err)

	// Release the lock
	err = c.lock.ReleaseLock(ctx, "", LOCK2)
	assert.Nil(t, err)

	// Acquire lock for the third time
	err = c.lock.AcquireLock(ctx, "", LOCK2, 3000, 1000)
	assert.Nil(t, err)

	c.lock.ReleaseLock(ctx, "", LOCK2)
}

func (c *LockFixture) TestReleaseLock(t *testing.T) {
	ctx := context.Background()

	// Acquire lock for the first time
	result, err := c.lock.TryAcquireLock(ctx, "", LOCK3, 3000)
	assert.Nil(t, err)
	assert.True(t, result)

	// Release the lock for the first time
	err = c.lock.ReleaseLock(ctx, "", LOCK3)
	assert.Nil(t, err)
	// Release the lock for the second time
	err = c.lock.ReleaseLock(ctx, "", LOCK3)
	assert.Nil(t, err)
}


test/lock/RedisLock_test.go
package test_lock

import (
	"context"
	"os"
	"testing"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	redislock "github.com/pip-services3-gox/pip-services3-redis-gox/lock"
	redisfixture "github.com/pip-services3-gox/pip-services3-redis-gox/test/fixture"
)

func TestRedisLock(t *testing.T) {
	var lock *redislock.RedisLock
	var fixture *redisfixture.LockFixture

	ctx := context.Background()

	host := os.Getenv("REDIS_SERVICE_HOST")
	if host == "" {
		host = "localhost"
	}

	port := os.Getenv("REDIS_SERVICE_PORT")
	if port == "" {
		port = "6379"
	}

	lock = redislock.NewRedisLock()

	config := cconf.NewConfigParamsFromTuples(
		"connection.host", host,
		"connection.port", port,
	)
	lock.Configure(ctx, config)
	fixture = redisfixture.NewLockFixture(lock)

	lock.Open(ctx, "")
	defer lock.Close(ctx, "")

	t.Run("Try Acquire Lock", fixture.TestTryAcquireLock)
	t.Run("Acquire Lock", fixture.TestAcquireLock)
	t.Run("Release Lock", fixture.TestReleaseLock)
}


test/fixture/CacheFixture.go
package test_fixture

import (
	"context"
	"encoding/json"
	"testing"
	"time"

	ccache "github.com/pip-services3-gox/pip-services3-components-gox/cache"
	"github.com/stretchr/testify/assert"
)

var (
	KEY1 string = "key1"
	KEY2 string = "key2"
	KEY3 string = "key3"
	KEY4 string = "key4"
	KEY5 string = "key5"
	KEY6 string = "key6"

	VALUE1 string         = "value1"
	VALUE2 map[string]any = map[string]any{"val": "value2"}
	VALUE3 time.Time      = time.Now()
	VALUE4 []int          = []int{1, 2, 3, 4}
	VALUE5 int            = 12345
	VALUE6 any            = nil
)

type CacheFixture struct {
	cache ccache.ICache[any]
}

func NewCacheFixture(cache ccache.ICache[any]) *CacheFixture {
	c := CacheFixture{}
	c.cache = cache
	return &c
}

func (c *CacheFixture) TestStoreAndRetrieve(t *testing.T) {
	ctx := context.Background()

	_, err := c.cache.Store(ctx, "", KEY1, VALUE1, 5000)
	assert.Nil(t, err)

	_, err = c.cache.Store(ctx, "", KEY2, VALUE2, 5000)
	assert.Nil(t, err)

	_, err = c.cache.Store(ctx, "", KEY3, VALUE3, 5000)
	assert.Nil(t, err)

	_, err = c.cache.Store(ctx, "", KEY4, VALUE4, 5000)
	assert.Nil(t, err)

	_, err = c.cache.Store(ctx, "", KEY5, VALUE5, 5000)
	assert.Nil(t, err)

	_, err = c.cache.Store(ctx, "", KEY6, VALUE6, 5000)
	assert.Nil(t, err)

	<-time.After(500 * time.Millisecond)

	val, err := c.cache.Retrieve(ctx, "", KEY1)
	assert.Nil(t, err)
	assert.NotNil(t, val)
	assert.Equal(t, VALUE1, val.(string))

	val, err = c.cache.Retrieve(ctx, "", KEY2)

	assert.Nil(t, err)
	assert.NotNil(t, val)
	assert.Equal(t, VALUE2, val.(map[string]any))

	val, err = c.cache.Retrieve(ctx, "", KEY3)
	assert.Nil(t, err)

	marshalVal3, err := json.Marshal(VALUE3)
	assert.Nil(t, err)

	assert.NotNil(t, val)
	assert.Equal(t, string(marshalVal3), ("\"" + val.(string) + "\""))

	val, err = c.cache.Retrieve(ctx, "", KEY4)

	assert.Nil(t, err)
	assert.NotNil(t, val)

	intArr := make([]int, 0, 4)

	for _, v := range val.([]any) {
		intArr = append(intArr, int(v.(float64)))
	}

	assert.Equal(t, VALUE4, intArr)

	val, err = c.cache.Retrieve(ctx, "", KEY5)

	assert.Nil(t, err)
	assert.NotNil(t, val)
	assert.Equal(t, VALUE5, int(val.(float64)))

	val, err = c.cache.Retrieve(ctx, "", KEY6)

	assert.Nil(t, err)
	assert.Nil(t, val)
}

func (c *CacheFixture) TestRetrieveExpired(t *testing.T) {
	ctx := context.Background()

	_, err := c.cache.Store(ctx, "", KEY1, VALUE1, 1000)
	assert.Nil(t, err)

	<-time.After(1500 * time.Millisecond)

	val, err := c.cache.Retrieve(ctx, "", KEY1)
	assert.Nil(t, err)
	assert.Nil(t, val)
}

func (c *CacheFixture) TestRemove(t *testing.T) {
	ctx := context.Background()

	_, err := c.cache.Store(ctx, "", KEY1, VALUE1, 1000)
	assert.Nil(t, err)

	err = c.cache.Remove(ctx, "", KEY1)
	assert.Nil(t, err)

	val, err := c.cache.Retrieve(ctx, "", KEY1)
	assert.Nil(t, err)
	assert.Nil(t, val)
}


lock/RedisLock.go
package lock

import (
	"context"
	"strconv"
	"time"

	"github.com/gomodule/redigo/redis"
	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cdata "github.com/pip-services3-gox/pip-services3-commons-gox/data"
	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"
	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	cauth "github.com/pip-services3-gox/pip-services3-components-gox/auth"
	ccon "github.com/pip-services3-gox/pip-services3-components-gox/connect"
	clock "github.com/pip-services3-gox/pip-services3-components-gox/lock"
)

/*
RedisLock are distributed lock that is implemented based on Redis in-memory database.

Configuration parameters:

  - connection(s):
    - discovery_key:         (optional) a key to retrieve the connection from IDiscovery
    - host:                  host name or IP address
    - port:                  port number
    - uri:                   resource URI or connection string with all parameters in it
  - credential(s):
    - store_key:             key to retrieve parameters from credential store
    - username:              user name (currently is not used)
    - password:              user password
  - options:
    - retrytimeout:         timeout in milliseconds to retry lock acquisition. (Default: 100)
    - retries:               number of retries (default: 3)
    - db_num:                database number in Redis  (default 0)

References:

- *:discovery:*:*:1.0        (optional) IDiscovery services to resolve connection
- *:credential-store:*:*:1.0 (optional) Credential stores to resolve credential

Example:
	ctx := context.Background()

    lock = NewRedisRedis();
    lock.Configure(ctx, cconf.NewConfigParamsFromTuples(
      "host", "localhost",
      "port", 6379,
    ));

    err = lock.Open(ctx, "123")
      ...

    result, err := lock.TryAcquireLock(ctx, "123", "key1", 3000)
    if result {
    	// Processing...
    }
    err = lock.ReleaseLock(ctx, "123", "key1")
    // Continue...
*/
type RedisLock struct {
	*clock.Lock
	connectionResolver *ccon.ConnectionResolver
	credentialResolver *cauth.CredentialResolver

	lockId  string
	timeout int
	retries int
	dbNum   int

	client redis.Conn
}

// NewRedisLock method are creates a new instance of this lock.
func NewRedisLock() *RedisLock {
	c := &RedisLock{
		connectionResolver: ccon.NewEmptyConnectionResolver(),
		credentialResolver: cauth.NewEmptyCredentialResolver(),
		lockId:             cdata.IdGenerator.NextLong(),
		timeout:            30000,
		retries:            3,
		dbNum:              0,
		client:             nil,
	}
	c.Lock = clock.InheritLock(c)
	return c
}

// Configure method are configures component by passing configuration parameters.
// Parameters:
//   - config    configuration parameters to be set.
func (c *RedisLock) Configure(ctx context.Context, config *cconf.ConfigParams) {
	c.connectionResolver.Configure(ctx, config)
	c.credentialResolver.Configure(ctx, config)

	c.timeout = config.GetAsIntegerWithDefault("options.timeout", c.timeout)
	c.retries = config.GetAsIntegerWithDefault("options.retries", c.retries)
	c.dbNum = config.GetAsIntegerWithDefault("options.db_num", c.dbNum)
	if c.dbNum > 15 || c.dbNum < 0 {
		c.dbNum = 0
	}
}

// SetReferences method are sets references to dependent components.
// Parameters:
//   - ctx context.Context
//   - references 	references to locate the component dependencies.
func (c *RedisLock) SetReferences(ctx context.Context, references cref.IReferences) {
	c.connectionResolver.SetReferences(ctx, references)
	c.credentialResolver.SetReferences(ctx, references)
}

// IsOpen method are checks if the component is opened.
// Returns true if the component has been opened and false otherwise.
func (c *RedisLock) IsOpen() bool {
	return c.client != nil
}

// Open method are opens the component.
// Parameters:
//  - ctx context.Context
// 	- correlationId 	(optional) transaction id to trace execution through call chain.
// Returns: error or nil no errors occured.
func (c *RedisLock) Open(ctx context.Context, correlationId string) error {
	var connection *ccon.ConnectionParams
	var credential *cauth.CredentialParams

	connection, err := c.connectionResolver.Resolve(correlationId)

	if err == nil && connection == nil {
		err = cerr.NewConfigError(correlationId, "NO_CONNECTION", "Connection is not configured")
		return err
	}

	credential, err = c.credentialResolver.Lookup(ctx, correlationId)
	if err != nil {
		return err
	}

	var url, host, port, password string
	var dialOpts []redis.DialOption = make([]redis.DialOption, 0)

	dialOpts = append(dialOpts, redis.DialConnectTimeout(time.Duration(c.timeout)*time.Millisecond))
	dialOpts = append(dialOpts, redis.DialDatabase(c.dbNum))

	if credential != nil {
		password = credential.Password()
		dialOpts = append(dialOpts, redis.DialPassword(password))
	}

	if connection.Uri() != "" {
		url = connection.Uri()
		c.client, err = redis.DialURL(url, dialOpts...)
	} else {
		host = connection.Host()
		if host == "" {
			host = "localhost"
		}
		port = strconv.FormatInt(int64(connection.Port()), 10)
		if port == "0" {
			port = "6379"
		}
		url = host + ":" + port
		c.client, err = redis.Dial("tcp", url, dialOpts...)
	}
	return err
}

// Close method are closes component and frees used resources.
// Parameters:
//  - ctx context.Context
//  - correlationId 	(optional) transaction id to trace execution through call chain.
// Retruns: error or nil no errors occured.
func (c *RedisLock) Close(ctx context.Context, correlationId string) error {
	if c.client != nil {
		err := c.client.Close()
		c.client = nil
		if err != nil {
			return err
		}
	}
	return nil
}

func (c *RedisLock) checkOpened(correlationId string) (state bool, err error) {
	if !c.IsOpen() {
		err = cerr.NewInvalidStateError(correlationId, "NOT_OPENED", "Connection is not opened")
		return false, err
	}

	return true, nil
}

// TryAcquireLock method are makes a single attempt to acquire a lock by its key.
// It returns immediately a positive or negative result.
// Parameters:
//  - ctx context.Context
//  - correlationId     (optional) transaction id to trace execution through call chain.
//  - key               a unique lock key to acquire.
//  - ttl               a lock timeout (time to live) in milliseconds.
// Returns: a lock result or error.
func (c *RedisLock) TryAcquireLock(ctx context.Context, correlationId string, key string, ttl int64) (result bool, err error) {
	state, err := c.checkOpened(correlationId)
	if !state {
		return false, err
	}

	res, err := redis.String(c.client.Do("SET", key, c.lockId, "NX", "PX", ttl))
	if err != nil && err == redis.ErrNil {
		return false, nil
	}
	return res == "OK", err
}

// ReleaseLock method are releases prevously acquired lock by its key.
//  - ctx context.Context
//  - correlationId     (optional) transaction id to trace execution through call chain.
//  - key               a unique lock key to release.
// Returns: error or nil for success.
func (c *RedisLock) ReleaseLock(ctx context.Context, correlationId string, key string) error {
	state, err := c.checkOpened(correlationId)
	if !state {
		return err
	}

	// Start transaction on key
	_, err = c.client.Do("WATCH", key)
	if err != nil {
		return err
	}

	// Read and check if lock is the same
	keyId, err := redis.String(c.client.Do("GET", key))
	if err != nil && err != redis.ErrNil {
		c.client.Do("UNWATCH")
		return err
	}
	// Remove the lock if it matches
	if keyId == c.lockId {
		c.client.Send("MULTI")
		c.client.Send("DEL", key)
		_, err = c.client.Do("EXEC")
	} else { // Cancel transaction if it doesn"t match
		_, err = c.client.Do("UNWATCH")
	}
	return err
}


cache/RedisCache.go
package persistence

import (
	"context"

	"strconv"
	"time"

	"github.com/go-redis/redis"
	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cconv "github.com/pip-services3-gox/pip-services3-commons-gox/convert"
	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"
	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	cauth "github.com/pip-services3-gox/pip-services3-components-gox/auth"
	ccon "github.com/pip-services3-gox/pip-services3-components-gox/connect"
	clog "github.com/pip-services3-gox/pip-services3-components-gox/log"
)

/*
Distributed cache that stores values in Redis in-memory database.

Configuration parameters:

  - connection(s):
    - discovery_key:         (optional) a key to retrieve the connection from IDiscovery
    - host:                  host name or IP address
    - port:                  port number
    - uri:                   resource URI or connection string with all parameters in it
  - credential(s):
    - store_key:             key to retrieve parameters from credential store
    - username:              user name (currently is not used)
    - password:              user password
  - options:
    - retries:               number of retries (default: 3)
    - timeout:               default caching timeout in milliseconds (default: 1 minute)
    - db_num:                database number in Redis  (default 0)
    - max_size:            	 maximum number of values stored in this cache (default: 1000)
    - cluster:            	 enable redis cluster

References:

- *:discovery:*:*:1.0        (optional) IDiscovery services to resolve connection
- *:credential-store:*:*:1.0 (optional) Credential stores to resolve credential

Example:
	ctx := context.Background()

    cache = NewRedisCache[string]();
    cache.Configure(ctx, cconf.NewConfigParamsFromTuples(
      "host", "localhost",
      "port", 6379,
    ));

    err = cache.Open(ctx, "123")
      ...

    ret, err := cache.Store(ctx, "123", "key1", []byte("ABC"))
    if err != nil {
    	...
    }

    res, err := cache.Retrieve(ctx, "123", "key1")
    value, _ := res.([]byte)
    fmt.Println(string(value))     // Result: "ABC"
*/
type RedisCache[T any] struct {
	connectionResolver *ccon.ConnectionResolver
	credentialResolver *cauth.CredentialResolver

	timeout   int
	retries   int
	dbNum     int
	isCluster bool

	client redis.UniversalClient
	logger clog.CompositeLogger

	convertor cconv.IJSONEngine[T]
}

// NewRedisCache method are creates a new instance of this cache.
func NewRedisCache[T any]() *RedisCache[T] {
	return &RedisCache[T]{
		logger:             *clog.NewCompositeLogger(),
		connectionResolver: ccon.NewEmptyConnectionResolver(),
		credentialResolver: cauth.NewEmptyCredentialResolver(),
		timeout:            30000,
		retries:            30000,
		dbNum:              3,
		convertor:          cconv.NewDefaultCustomTypeJsonConvertor[T](),
	}
}

// Configure method are configures component by passing configuration parameters.
// 	 - ctx context.Context
//   - config    configuration parameters to be set.
func (c *RedisCache[T]) Configure(ctx context.Context, config *cconf.ConfigParams) {
	c.connectionResolver.Configure(ctx, config)
	c.credentialResolver.Configure(ctx, config)
	c.logger.Configure(ctx, config)

	c.timeout = config.GetAsIntegerWithDefault("options.timeout", c.timeout)
	c.retries = config.GetAsIntegerWithDefault("options.retries", c.retries)
	c.dbNum = config.GetAsIntegerWithDefault("options.db_num", c.dbNum)
	if c.dbNum > 15 || c.dbNum < 0 {
		c.dbNum = 0
	}
	c.isCluster = config.GetAsBooleanWithDefault("options.cluster", c.isCluster)
}

// Sets references to dependent components.
//	 - ctx context.Context
//   - references 	references to locate the component dependencies.
func (c *RedisCache[T]) SetReferences(ctx context.Context, references cref.IReferences) {
	c.connectionResolver.SetReferences(ctx, references)
	c.credentialResolver.SetReferences(ctx, references)
	c.logger.SetReferences(ctx, references)
}

// Checks if the component is opened.
// Returns true if the component has been opened and false otherwise.
func (c *RedisCache[T]) IsOpen() bool {
	return c.client != nil
}

// Open method are opens the component.
// Parameters:
//  - ctx context.Context
//  - correlationId 	(optional) transaction id to trace execution through call chain.
// Returns: error or nil no errors occured.
func (c *RedisCache[T]) Open(ctx context.Context, correlationId string) error {
	var (
		connection *ccon.ConnectionParams
		credential *cauth.CredentialParams
		options    redis.Options
	)

	connection, err := c.connectionResolver.Resolve(correlationId)

	if err == nil && connection == nil {
		err = cerr.NewConfigError(correlationId, "NO_CONNECTION", "Connection is not configured")
		return err
	}

	credential, err = c.credentialResolver.Lookup(ctx, correlationId)

	if err != nil {
		return err
	}

	options.DialTimeout = time.Duration(c.timeout) * time.Millisecond
	options.DB = c.dbNum
	options.MaxRetries = c.retries

	if credential != nil {
		options.Password = credential.Password()
	}

	if connection.Uri() != "" {
		options.Addr = connection.Uri()
	} else {
		host := connection.Host()
		if host == "" {
			host = "localhost"
		}
		port := strconv.FormatInt(int64(connection.Port()), 10)
		if port == "0" {
			port = "6379"
		}
		options.Addr = host + ":" + port
	}
	if c.isCluster {
		c.client = redis.NewClusterClient(&redis.ClusterOptions{
			OnNewNode: func(client *redis.Client) {
				client = redis.NewClient(&options)
			},
			Addrs:       append([]string{}, options.Addr),
			Password:    options.Password,
			DialTimeout: options.DialTimeout,
		})
	} else {
		c.client = redis.NewClient(&options)
	}
	_, err = c.client.Ping().Result()
	return err
}

// Close method are closes component and frees used resources.
// Parameters:
//   - ctx context.Context
//   - correlationId 	(optional) transaction id to trace execution through call chain.
// Retruns: error or nil no errors occured.
func (c *RedisCache[T]) Close(ctx context.Context, correlationId string) error {
	if c.client != nil {
		err := c.client.Close()
		c.client = nil
		if err != nil {
			return err
		}
	}
	return nil
}

func (c *RedisCache[T]) checkOpened(correlationId string) (state bool, err error) {
	if !c.IsOpen() {
		err = cerr.NewInvalidStateError(correlationId, "NOT_OPENED", "Connection is not opened")
		return false, err
	}

	return true, nil
}

// Retrieve method are retrieves cached value from the cache using its key.
// If value is missing in the cache or expired it returns nil.
// Parameters:
//   - ctx context.Context
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
//  Retruns: cached value or error.
func (c *RedisCache[T]) Retrieve(ctx context.Context, correlationId string, key string) (value T, err error) {
	var defaultValue T

	if state, err := c.checkOpened(correlationId); !state {
		return defaultValue, err
	}

	item, err := c.client.Get(key).Bytes()

	if err != nil {
		if err == redis.Nil {
			return defaultValue, nil
		}
		return defaultValue, err
	}

	if item != nil {
		val, err := c.convertor.FromJson(string(item))
		if err != nil {
			return defaultValue, err
		}
		return val, nil
	}

	return defaultValue, nil
}

// Store method are stores value in the cache with expiration time.
// Parameters:
//   - ctx context.Context
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
//   - value             a value to store.
//   - timeout           expiration timeout in milliseconds.
// Retruns error or nil for success
func (c *RedisCache[T]) Store(ctx context.Context, correlationId string, key string, value T, timeout int64) (result T, err error) {
	state, err := c.checkOpened(correlationId)

	var defaultValue T

	if !state {
		return defaultValue, err
	}

	jsonVal, err := c.convertor.ToJson(value)
	if err != nil {
		return defaultValue, err
	}
	tmout := time.Duration(timeout) * time.Millisecond
	return value, c.client.Set(key, jsonVal, tmout).Err()
}

// Removes a value from the cache by its key.
// Parameters:
//   - ctx context.Context
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
// Returns: error or nil for success
func (c *RedisCache[T]) Remove(ctx context.Context, correlationId string, key string) error {
	state, err := c.checkOpened(correlationId)
	if !state {
		return err
	}
	return c.client.Del(key).Err()
}

// Contains check is value stores
// Parameters:
//   - ctx context.Context
//   - correlationId     (optional) transaction id to trace execution through call chain.
//   - key               a unique value key.
func (c *RedisCache[T]) Contains(ctx context.Context, correlationId string, key string) bool {
	state, err := c.checkOpened(correlationId)
	if !state {
		c.logger.Error(ctx, correlationId, err, "Connection is not opened")
		return false
	}
	return c.client.Exists(key).Val() > 0
}
